src/foam/nanos/iban/ValidationIBAN.js

```diff
@@ -150,98 +154,274 @@ foam.CLASS({
-        var c = iban.charAt(l);
-        if ( c >= 'A' && c <= 'Z' ) {
-          iban = iban.substring(0, l) + (10 + ( c.charCodeAt(0) - 'A'.charCodeAt(0) )) + iban.substring(l+1)
+        if ( iban.length() != format.length() + 2 )
+          throw new ValidationException(LENGTH_MISMATCHED_1 + (format.length() + 2) + LENGTH_MISMATCHED_2 + iban.length());
+
+        for ( int i = 0 ; i < format.length() ; i++ ) {
+          if ( ! validateChar(format.charAt(i), iban.charAt(i+2) )) throw new ValidationException(INVALID_CHARACTER + (i+2));
         }
-      }
-      return iban;
-    },
 
-    function mod(num, a) {
-      /** IBANs may be too long to represent as integers, so implement a string-based mod function. **/
-      var ret = 0;
+        String num = toNumber(iban);
+        int checksum = mod(num, 97);
 
-      // One by one process all digits of 'num'
-      for ( var i = 0 ; i < num.length ; i++ )
-        ret = (ret*10 + (num.charAt(i) - '0')) % a;
+        if ( checksum != 1 ) throw new ValidationException(INVALID_CHECKSUM + checksum);
 
-      return ret;
+      `
+    },
+    {
+      name: 'setChecksum',
+      type: 'String',
+      args: [
+        {
+          name: 'iban',
+          type: 'String'
+        }
+      ],
+      code: function(iban) {
+        iban = iban.substring(0,2) + '00' + iban.substring(4, 200);
+        let num = this.toNumber(iban);
+        let checksum = this.mod(num, 97);
+        let desiredChecksum = 98 - checksum;
+        iban = iban.substring(0,2) + ('' + desiredChecksum).padStart(2, '0') + iban.substring(4);
+        return iban;
+      },
+      javaCode: `
+        iban = iban.substring(0,2) + "00" + iban.substring(4, 200);
+        String num = toNumber(iban);
+        int checksum = mod(num, 97);
+        int desiredChecksum = 98 - checksum;
+        iban = iban.substring(0,2) + ("" + String.format("%2s", "" + desiredChecksum).replace(' ', '0') + iban.substring(4));
+        return iban;
+      `
+    },
+    {
+      name: 'toNumber',
+      type: 'String',
+      args: [
+        {
+          name: 'iban',
+          type: 'String'
+        }
+      ],
+      code: function(iban) {
+        iban = iban.substring(4) + iban.substring(0, 4);
+        let c = "";
+        for ( let l = iban.length - 1; l >= 0 ; l-- ) {
+          c = iban.charAt(l);
+          console.log(!!c);
+          if ( c >= 'A' && c <= 'Z' ) {
+            iban = iban.substring(0, l) + (10 + ( c.charCodeAt(0) - 'A'.charCodeAt(0) )) + iban.substring(l+1);
+          }
+        }
+        return iban;
+      },
+      javaCode: `
+        iban = iban.substring(4) + iban.substring(0, 4);
+        for ( int l = iban.length() - 1 ; l >= 0 ; l--) {
+          char c = iban.charAt(l);
+          if ( c >= 'A' && c <= 'Z' ) {
+           iban = iban.substring(0, l) + (10 + (Character.codePointAt(Character.toString(c), 0) - Character.codePointAt(Character.toString('A'), 0))) + iban.substring(l+1);
```

Comment:
Builds a new string for each character which will be slow. Would be better to create a char[] at the beginning.